c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -227,12 +227,18 @@ impl<'tcx> TypeChecker<'tcx, '_> {
         match stmt.kind {
             StatementKind::Assign(ref x) => {
                 let (pl, ref rv) = **x;
+                if matches!(rv, Rvalue::Cast(_, Operand::Copy(p) | Operand::Move(p), _) if self.acx.special_casts.contains_key(&p) || self.acx.special_casts.contains_key(&pl))
```

Comment:
I think this part, specifically the `matches!(rv, Rvalue::Cast(_, Operand::Copy(p) | Operand::Move(p), _) if self.acx.special_casts.contains_key(&p)` can be extracted to a `SpecialCasts` method as its repeated above as well.